### PR TITLE
Task 8: Maven project, servlet to handle ads

### PR DESCRIPTION
**Pull request by marybash**
* created a maven project based on the application from task 5;
* created a Rest API: a servlet to handle ads;
* implemented doPost, doGet, doDelete and doPut methods;
* tested the created servlet;
* added UI files;

### CODE REVIEW

File: ProfitSite/src/main/java/Profit/AdFilters.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        return dateFrom;
+    }
+
+    public Date getDDateFrom() {
+        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd");
+        Date date = new Date();
+        try {
+            date = sdf.parse(dateFrom);
+        } catch (ParseException e) {
+            e.printStackTrace();
```
Comment (smiakchilo) [inline ProfitSite/src/main/java/Profit/AdFilters.java]: Consider using a logger instead. Printing to stdout/stderr is not used in client-server apps

File: ProfitSite/src/main/java/Profit/AdFilters.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        }
+        return date;
+    }
+
+    public String getDateTo() {
+        return dateTo;
+    }
+
+    public Date getDDateTo() {
```
Comment (smiakchilo) [inline ProfitSite/src/main/java/Profit/AdFilters.java]: Oh, this method name is tricky to understand. Please consider whether you need two different getters for the same value at all?

File: ProfitSite/src/main/java/Profit/AdItem.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
+package Profit;
+
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.*;
+
+public class AdItem {
+
+    private String id;
```
Comment (smiakchilo) [inline ProfitSite/src/main/java/Profit/AdItem.java]: Generally it's a good practice to add `final` to declarations of fields that you initialize only once (like in a constructor)

File: ProfitSite/src/main/java/Profit/AdItem.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    public String getDescription() {
+        return description;
+    }
+
+    public String getCreatedAt() {
+        return createdAt;
+    }
+
+    public Date getDateCreatedAt() {
+        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd");
```
Comment (smiakchilo) [inline ProfitSite/src/main/java/Profit/AdItem.java]: Even if you decide to leave this method be, it nearly copies the one from the previous class, which is no good. `getDateVa;lidUntil` is yet another copypase. In cases like this, you need to either create a common superclass for both classes sharing the same code or create a "utility" method that would be accessible from both classes

File: ProfitSite/src/main/java/Profit/AdFilters.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
+package Profit;
```
Comment (smiakchilo) [inline ProfitSite/src/main/java/Profit/AdFilters.java]: Package names must be all in lowercase by convention

File: ProfitSite/src/main/java/Profit/AdList.java [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+    public List<AdItem> getPage (AdFilters filters) {
+        String skip = filters.getSkip();
+        String top = filters.getTop();
+        boolean usedFilter = false;
+        if (skip == null) {
+            skip = "0";
+        }
+        if (top == null) {
+            top = "10";
```
Comment (smiakchilo) [inline ProfitSite/src/main/java/Profit/AdList.java]: Replace "magic numbers" with constants

File: ProfitSite/src/main/java/Profit/AdList.java [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+public class AdList {
+    private List<AdItem> ads;
+
+    private static class SortByDescendingDate implements Comparator<AdItem> {
+        public int compare(AdItem a, AdItem b) {
+            SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd");
+            Date date1 = new Date();
+            Date date2 = new Date();
+            try {
+                date1 = sdf.parse(a.getCreatedAt());
```
Comment (smiakchilo) [inline ProfitSite/src/main/java/Profit/AdList.java]: Why parsing again? You could use `getDateCreateAt()` )
But I strictly recommend that you just store dates as `Date`s from the start, and not at all strings

File: ProfitSite/src/main/java/Profit/AdList.java [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+        String top = filters.getTop();
+        boolean usedFilter = false;
+        if (skip == null) {
+            skip = "0";
+        }
+        if (top == null) {
+            top = "10";
+        }
+        List<AdItem> returningAds = ads;
+        if (filters.getDateFrom() != null) {
```
Comment (smiakchilo) [inline ProfitSite/src/main/java/Profit/AdList.java]: This method is too long and needs splitting

File: ProfitSite/src/main/java/Profit/AdList.java [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+        for(AdItem ad : ads) {
+            if(ad.getId().equals(id)) {
+                return ad;
+            }
+        }
+        return null;
+    }
+
+    public static boolean validate (AdItem ad) {
+        return ad.getId() != null && ad.getId().length() >= 1
```
Comment (smiakchilo) [inline ProfitSite/src/main/java/Profit/AdList.java]: Is `ad` guaranteed to never be null?

**Review by smiakchilo — COMMENTED** [review]
My fault but the PR is named "Maven project.. etc" and yet I can't find the POM file in project root) Did your project compile well?)

**Comment by marybash** [comment]
> My fault but the PR is named "Maven project.. etc" and yet I can't find the POM file in project root) Did your project compile well?)

I'm sorry, something went wrong and pom.xml file hasn't been attached, but now all's ok)

**Review by smiakchilo — APPROVED** [review]

**Pull request merged**